refacto: early exit estimate fee in runtime

tdelabro opened the pull request:
close #1315

EvolveArt commented on crates/pallets/starknet/src/utils.rs:
```diff
@@ -48,3 +48,50 @@ pub fn execute_txs_and_rollback<T: pallet::Config>(
+            // simple with the error handling
+            match result {
+                Ok(tx_info) if !tx_info.is_reverted() => execution_infos.push(tx_info),
+                _ => break,
```
I don't think we should simply break but rather return early with an error in the rollback no?

EvolveArt replied:
Also we might want to simply use the transactional [macro](https://paritytech.github.io/polkadot-sdk/master/frame_support/attr.transactional.html) wdyt ?

tdelabro replied:
> I don't think we should simply break but rather return early with an error in the rollback no?

We need to execute this line of code `storage::TransactionOutcome::Rollback(Result::<_, DispatchError>::Ok(()))`. Which says "I successfully executed what I needed to execut

tdelabro replied:
> Also we might want to simply use the transactional [macro](https://paritytech.github.io/polkadot-sdk/master/frame_support/attr.transactional.html) wdyt ?

For transactional function we need to return an error in order to revert storage changes. But we also need to get a vec execution infos out of this. Both can't be done

tdelabro replied:
> For transactional function we need to return an error in order to revert storage changes. But we also need to get a vec execution infos out of this. Both can't be done


I was wrong! It is doable. And it's in the new commits

EvolveArt commented on crates/pallets/starknet/src/utils.rs:
```diff
@@ -48,3 +48,50 @@ pub fn execute_txs_and_rollback<T: pallet::Config>(
+    })
+    .map_err(|_| Error::<T>::FailedToCreateATransactionalStorageExecution)?;
+
+    // All transation were successfully executed
```
```suggestion
    // All transactions were successfully executed
```

apoorvsadana commented on crates/pallets/starknet/src/lib.rs:
```diff
@@ -1079,31 +1081,27 @@ impl<T: Config> Pallet<T> {
-        }
-        Ok(results)
+        )?
+        .map_err(|_| DispatchError::from(Error::<T>::TransactionExecutionFailed))?;
```
can we log the error

tdelabro replied:
we would be logging `PlaceHolderErrorTypeForFailedStarknetExecution`. Nothing really intresting

apoorvsadana commented on crates/pallets/starknet/src/utils.rs:
```diff
@@ -48,3 +48,50 @@ pub fn execute_txs_and_rollback<T: pallet::Config>(
+            // simple with the error handling
+            match result {
+                Ok(tx_info) if !tx_info.is_reverted() => execution_infos.push(tx_info),
+                _ => break,
```
can we log the error here?

tdelabro commented:
@apoorvsadana I made some changes. More PR around here is coming due to deoxys implementing the rpc v0.6.0 that have more trace_api method.
I removed the db transactional thing. We now have to be extra careful but it simplifies the code for everybody.
And I moved all this logic to its own module, with clear warnings.

Could you review it again?

apoorvsadana commented on crates/pallets/starknet/src/simulations.rs:
```diff
@@ -0,0 +1,134 @@
+            .into_iter()
+            .map(|tx| {
+                execution_config.set_offset_version(tx.offset_version());
+                match tx {
+                    UserTransaction::Declare(tx, contract_class) => tx
+                        .try_into_executable::<T::SystemHash>(chain_id, contract_class.clone(), tx.offset_version())
+                        .and_then(|exec| {
+                            exec.execute(&mut BlockifierStateAdapter::<T>::default(), &block_context, &execution_config)
+                        }),
+                    UserTransaction::DeployAccount(tx) => {
+                        let executable = tx.into_executable::<T::SystemHash>(chain_id, tx.offset_version());
+                        executable.execute(
+                            &mut BlockifierStateAdapter::<T>::default(),
+                            &block_context,
+                            &execution_config,
+                        )
+                    }
+                    UserTransaction::Invoke(tx) => {
+                        let executable = tx.into_executable::<T::SystemHash>(chain_id, tx.offset_version());
+                        executable.execute(
+                            &mut BlockifierStateAdapter::<T>::default(),
+                            &block_context,
+                            &execution_config,
+                        )
+                    }
+                }
```
let's move the common code to get exeuction_info to a function? also, i didn't understand, in this case how are the changes reverted?

tdelabro replied:
They are not. Except if you call from the RuntimeAPI, then everything you do is reverted anyway.
That's why I added those big warning comments. And we have to be extra careful not to let anyone write code that would use those from inside an extrinsic

tdelabro replied:
> let's move the common code to get exeuction_info to a function?

Done
